common/src/main/scala/explore/common/RemoteSyncUndoableF.scala

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2016-2022 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package explore.common
+
+import cats.MonadError
+import crystal.react.View
+import crystal.react.implicits._
+import explore.undo.UndoContext
+import japgolly.scalajs.react.Reusability
+import japgolly.scalajs.react.callback.Callback
+import japgolly.scalajs.react.util.Effect
+import monocle.Lens
+import monocle.Optional
+import monocle.Prism
+import org.typelevel.log4cats.Logger
+
+/**
+ * Provides a `View` into a model with the ability to perform changes remotely whenever the `View`
+ * is modified. Additionally, operations are undoable.
+ *
+ * `A` is the model type, whereas `T` is the type of a delta structure to be sent to the remote
+ * server with only the necessary changes. For example, it can be one of the `*Input` types of the
+ * API.
+ *
+ * We can `zoom` into a property of the model while providing a parallel way to zoom into the delta
+ * structure. Potential modifications to the the delta structure are stored in a coyoneda-like
+ * fashion, but are not executed until the model is modified.
+ *
+ * The idea is to keep code to edit the model as simple as possible, while providing a way to
+ * compose components that edit different parts of the same model object in an undo context.
+ */
+trait RemoteSyncUndoableF[F[_], B, S] { self =>
+  protected type _A // Base model type
+  protected type _T // Base delta structure type
+
+  protected val _undoCtx: UndoContext[_A] // Base model instance within an `UndoContext`
+  protected val _remoteBaseInput: _T      // Base delta structure instance
+  protected val _onMod: _T => F[Unit]     // Effect to send the delta structure to the remote server
+
+  protected val _modelGet: _A => B               // Model drill-down getter function
+  protected val _modelMod: (B => B) => _A => _A  // Model drill-down modifier function
+  protected val _remoteMod: (S => S) => _T => _T // Delta structure drill-down modifier function
+
+  /** Get the value of the model. */
+  def get: B = _modelGet(_undoCtx.model.get)
+
+  /**
+   * Build an undoable `View` at the current level, specifying a function that modifies the delta
+   * structure so that model changes are added to the structure.
+   *
+   * The `View` should be built at the last possible moment in order to maximize the granularity of
+   * undo operations and remote changes.
+   */
+  def viewMod(toInput: B => S => S)(implicit
+    F:                 MonadError[F, Throwable],
+    dispatcher:        Effect.Dispatch[F],
+    logger:            Logger[F]
+  ): View[B] =
+    _undoCtx
+      .undoableView(_modelGet, _modelMod)
+      .withOnMod(value =>
+        Callback.log(_remoteMod(toInput(value))(_remoteBaseInput)) >>
```

Comment:
Was this for development. If not, should we log more information?

Reply:
Good catch. Yes, it was for debugging, I removed it now.